velox/expression/SignatureBinder.cpp

```diff
@@ -122,90 +133,114 @@ bool SignatureBinder::tryBind() {
   return true;
 }
 
+bool SignatureBinder::checkOrSetIntegerParameter(
+    const std::string& parameterName,
+    int value) {
+  auto it = integerParameters_.find(parameterName);
+  // Return false if the parameter is not found.
+  if (it == integerParameters_.end()) {
+    return false;
+  }
+  // Return false if the parameter is found with a different value.
+  if (it->second.has_value() && it->second.value() != value) {
+    return false;
+  }
+  it->second = value;
+  return true;
+}
+
+bool SignatureBinder::tryBindIntegerParameters(
+    const std::vector<exec::TypeSignature>& parameters,
+    const TypePtr& actualType) {
+  // Decimal types
+  if (actualType->isShortDecimal() || actualType->isLongDecimal()) {
+    VELOX_CHECK_EQ(parameters.size(), 2);
+    const auto& [precision, scale] = getDecimalPrecisionScale(*actualType);
+    return checkOrSetIntegerParameter(parameters[0].baseName(), precision) &&
+        checkOrSetIntegerParameter(parameters[1].baseName(), scale);
+  }
+  return false;
+}
+
 bool SignatureBinder::tryBind(
     const exec::TypeSignature& typeSignature,
     const TypePtr& actualType) {
   if (isAny(typeSignature)) {
     return true;
   }
-
-  auto it = bindings_.find(typeSignature.baseType());
-  if (it == bindings_.end()) {
-    // concrete type
-    auto typeName = boost::algorithm::to_upper_copy(typeSignature.baseType());
-    if (isDecimalName(typeName)) {
-      VELOX_USER_FAIL("Use 'DECIMAL' in the signature.");
-    }
-    if (isDecimalKind(actualType->kind()) && isCommonDecimalName(typeName)) {
-      const auto& variables = typeSignature.variables();
-      VELOX_CHECK_EQ(variables.size(), 2);
-      const auto& [precision, scale] = getDecimalPrecisionScale(*actualType);
-      variables_.emplace(variables[0], precision);
-      variables_.emplace(variables[1], scale);
-      return true;
-    }
+  const auto baseName = typeSignature.baseName();
+  if (isConcreteType(typeParameters_, integerParameters_, baseName)) {
+    auto typeName = boost::algorithm::to_upper_copy(baseName);
 
     if (typeName != actualType->kindName()) {
-      return false;
+      if (!(isCommonDecimalName(typeName) &&
+            (actualType->isLongDecimal() || actualType->isShortDecimal()))) {
+        return false;
+      }
     }
 
     const auto& params = typeSignature.parameters();
-    if (params.size() != actualType->size()) {
-      return false;
+    // Integer parameters have to be resolved here.
+    if (params.size() > 0 &&
+        integerParameters_.count(params[0].baseName()) != 0) {
```

Comment:
Maybe add a comment that we assume integer parameters start from the first position if exist.

Reply:
done